arrows/vxl/average_frames.cxx

```diff
@@ -0,0 +1,747 @@
+
+  // Internal parameters/settings
+  std::string type;
+  unsigned window_size;
+  double exp_weight;
+  bool round;
+  bool output_variance;
+  double variance_scale;
+
+  // The actual frame averager
+  frame_averager_byte_sptr byte_averager;
+  frame_averager_float_sptr float_averager;
+
+  // Load model, special optimizations are in place for the byte case
+  void
+  load_model( bool is_byte = true )
+  {
+    if( ( is_byte && byte_averager ) || ( !is_byte && float_averager ) )
+    {
+      return;
+    }
+
+    if( type == "window" )
+    {
+      if( is_byte )
+      {
+        byte_averager.reset(
+            new windowed_frame_averager< vxl_byte >( round, window_size ) );
+      }
+      else
+      {
+        float_averager.reset(
+            new windowed_frame_averager< double >( round, window_size ) );
+      }
+    }
+    else if( type == "cumulative" )
+    {
+      if( is_byte )
+      {
+        byte_averager.reset(
+            new cumulative_frame_averager< vxl_byte >( round ) );
+      }
+      else
+      {
+        float_averager.reset(
+            new cumulative_frame_averager< double >( round ) );
+      }
+    }
+    else if( type == "exponential" )
+    {
+      if( exp_weight <= 0 || exp_weight >= 1 )
+      {
+        throw std::runtime_error( "Invalid exponential averaging coefficient!" );
+      }
+
+      if( is_byte )
+      {
+        byte_averager.reset(
+            new exponential_frame_averager< vxl_byte >( round, exp_weight ) );
+      }
+      else
+      {
+        float_averager.reset(
+            new exponential_frame_averager< double >( round, exp_weight ) );
+      }
+    }
+    else
+    {
+      throw std::runtime_error( "Invalid averaging type!" );
+    }
+  }
+
+  // Compute the updated average with the current frame
+  // return the average or the variance
+  kwiver::vital::image_container_sptr
+  process_frame( vil_image_view< double > input )
+  {
+    load_model( false );
+
+    if( !output_variance )
+    {
+      vil_image_view< double > output;
+      float_averager->process_frame( input, output );
+      return std::make_shared< vxl::image_container >( output );
+    }
+    else
+    {
+      vil_image_view< double > tmp, output;
+      float_averager->process_frame( input, tmp, output );
+      return std::make_shared< vxl::image_container >( output );
+    }
+  }
+};
+
+// ----------------------------------------------------------------------------
+average_frames
+::average_frames()
+  : d( new priv() )
+{
+  attach_logger( "arrows.vxl.average_frames" );
+}
+
+// ----------------------------------------------------------------------------
+average_frames
+::~average_frames()
+{
+}
+
+// ----------------------------------------------------------------------------
+vital::config_block_sptr
+average_frames
+::get_configuration() const
+{
+  // get base config from base class
+  vital::config_block_sptr config = algorithm::get_configuration();
+
+  config->set_value( "type", d->type,
```

Comment:
If you feel like it, you can use the enum-converter approach here. If not, that's ok.

Reply:
Added, thanks.